@craigspaeth Image set

kanaabe opened the pull request:
Closes https://github.com/artsy/positron/issues/703 https://github.com/artsy/positron/issues/692 https://github.com/artsy/positron/issues/615 https://github.com/artsy/positron/issues/704

I apologies for the size of this PR! There are a couple more visual bugs I want to fix before merging but the main functionality is complete. 

![image_set](https://cloud.githubusercontent.com/assets/2236794/14332874/ac1bdf4c-fc19-11e5-8daf-81ed2a827135.gif)
Preview (4 should be centered in the svg): 
![image](https://cloud.githubusercontent.com/assets/2236794/14332890/cafe0336-fc19-11e5-8294-6cdb7d7fc3a5.png)

craigspaeth commented:
:-o this is super cool!

craigspaeth commented on api/apps/articles/model/schema.coffee:
```diff
@@ -74,20 +90,7 @@ fullscreenSection = (->
-            name: @string().allow('', null)
-            slug: @string().allow('', null)
-      ).allow(null).default([])
+      artworks: @array().items(denormalizedArtwork).allow(null).default([])
```
:+1:

craigspaeth commented on client/apps/edit/components/section_image_set/index.coffee:
```diff
@@ -0,0 +1,263 @@
+    totalPixels = 0.0
+    usedPixels = 0.0
+    $('.esis-preview-container .esis-preview-image').each (i, value) ->
+      _.defer ->
```
Good ol defer to the resuce.

craigspaeth commented on client/apps/edit/components/section_image_set/index.coffee:
```diff
@@ -0,0 +1,263 @@
+{ div, section, h1, h2, span, img, header, input, nav, a, button, p, ul, li, strong } = React.DOM
+{ crop, resize, fill } = embedly = require('embedly-view-helpers')(sd.EMBEDLY_KEY)
+
+module.exports = React.createClass
```
:clap: Nice work on this component! It's a little long in line numbers, but not _too_ long yet. If this grows in size I'd encourage trying to find way to break it down into even smaller parts.

craigspaeth commented on client/apps/edit/components/section_image_set/index.coffee:
```diff
@@ -0,0 +1,263 @@
+  onSelect: (e, selected) ->
+    new Artwork(id: selected.id).fetch
+      success: (artwork) =>
+        @state.images.push artwork.denormalized()
```
You shouldn't be mutating state directly here. Make sure to always use `setState` instead. e.g.

````coffeescript
newImages = @state.images.concat [artwork.denormalized()]
@setState images: newImages
````

craigspaeth commented on client/apps/edit/components/section_image_set/index.coffee:
```diff
@@ -0,0 +1,263 @@
+        image = new Image()
+        image.src = src
+        image.onload = =>
+          @state.images.push url: src, type: 'image'
```
Ditto as above

craigspaeth commented on client/apps/edit/components/section_image_set/index.coffee:
```diff
@@ -0,0 +1,263 @@
+          @onStateChange()
+
+  removeItem: (item) -> =>
+    @state.images = _.without @state.images, item
```
:eyes:

craigspaeth commented on client/apps/edit/components/section_image_set/index.coffee:
```diff
@@ -0,0 +1,263 @@
+      success: (artwork) =>
+        @refs.byUrl.setState loading: false, errorMessage: ''
+        $(@refs.byUrl.getDOMNode()).val ''
+        @state.images.push artwork.denormalized()
```
:eye

craigspaeth commented on client/apps/edit/components/section_image_set/index.coffee:
```diff
@@ -0,0 +1,263 @@
+  onStateChange: ->
+    @props.section.set images: @state.images if @state.images.length > 0
+    @showPreviewImages()
+    @forceUpdate()
```
You also should have to use `forceUpdate` unless you're hooking into something un-react-like (e.g. events on Backbone models). I bet if you refactor the state mutations into `setState` you'll notice you won't need this.

Also if `onStateChange` is meant to run after every "re-render" than I think a [lifecycle callback](https://facebook.github.io/react/docs/component-specs.html#updating-componentdidupdate) would be more appropriate.

craigspaeth commented on client/apps/edit/components/section_image_set/index.styl:
```diff
@@ -0,0 +1,263 @@
+  height 40px
+  display inline-block
+
+[data-useragent*=Firefox] .esis-nav
```
:+1:

craigspaeth commented on client/apps/edit/components/section_image_set/index.styl:
```diff
@@ -0,0 +1,263 @@
+  position relative
+  margin-top 20px
+
+.edit-section-image-set .edit-section-controls
```
Great namespacing/stylus work! Likewise with the line numbers here—not _too_ long yet, but if this thing grows also consider breaking down the stylesheets.

craigspaeth commented on client/apps/edit/components/section_image_set/test/index.coffee:
```diff
@@ -0,0 +1,98 @@
+{ div } = React.DOM
+fixtures = require '../../../../../../test/helpers/fixtures'
+
+describe 'SectionImageSet', ->
```
:heart: the tests!

craigspaeth commented on client/models/artwork.coffee:
```diff
@@ -17,3 +17,39 @@ module.exports = class Artwork extends Backbone.Model
   defaultImage: ->
     new AdditionalImages(@get('images')).default()
+
+  denormalized: ->
```
:+1: :+1:

craigspaeth commented on client/models/artwork.coffee:
```diff
@@ -17,3 +17,39 @@ module.exports = class Artwork extends Backbone.Model
+      name: @getArtistName()
+      slug: @get('artist')?.id
+
+  getArtistName: ->
```
:+1:

Although these might be something we want to [share across Backbone apps](https://github.com/artsy/artsy-backbone-mixins/tree/master/lib), or leverage Metaphysics for. Totally cool with this implementation for now though.

craigspaeth commented:
Looking good! Mainly concerned about the `state` mutations.

Also might want to consider moving the denormalizing artwork logic to the api-side and just have the client send down artwork ids. You could probably leverage some of the api-side code you wrote before in this case, and it would make it friendly for Orta when he writes a native iPad Writer app :stuck_out_tongue:

kanaabe commented:
Thanks for the comments @craigspaeth!

kanaabe commented:
I think the biggest opportunity after this PR is to make the image and artwork components pluggable into the image_set component. It's admittedly a lot of repeated code in various places. Anyway, I think this one is good to go now, I've changed the mutations and hooked into the `componentWillUpdate` lifecycle cb and removed `forceUpdate()` (you're right we totally didn't need this!)

craigspaeth commented:
w00t w00t! Nice work! :confetti_ball:
